lib/rust/parser/src/syntax/tree.rs

```diff
@@ -413,9 +465,35 @@ impl<'s> span::Builder<'s> for MultiSegmentAppSegment<'s> {
 
 /// Join two nodes with a new node appropriate for their types.
 ///
-/// For most input types, this simply constructs an `App`; however, for some block type operands
+/// For most input types, this simply constructs an `App`; however, for some operand types
 /// application has special semantics.
-pub fn apply<'s>(func: Tree<'s>, mut arg: Tree<'s>) -> Tree<'s> {
+pub fn apply<'s>(mut func: Tree<'s>, mut arg: Tree<'s>) -> Tree<'s> {
+    match &mut *func.variant {
+        Variant::OprApp(app)
+        if let Ok(opr) = &app.opr && !opr.can_form_section && app.rhs.is_none() => {
+            app.rhs = Some(arg);
+            return func;
+        }
+        Variant::TextLiteral(lhs) if let Variant::TextLiteral(rhs) = &mut *arg.variant
+                && lhs.close_quote.is_none() && rhs.open_quote.is_none() => {
+            match rhs.elements.first_mut() {
+                Some(TextElement::Section { text }) => text.left_offset = arg.span.left_offset,
+                Some(TextElement::Escape { backslash }) =>
+                    backslash.left_offset = arg.span.left_offset,
+                None => (),
+            }
+            lhs.elements.append(&mut rhs.elements);
+            lhs.close_quote = rhs.close_quote.take();
+            if rhs.trim != VisibleOffset(0) && (lhs.trim == VisibleOffset(0) || rhs.trim < lhs.trim) {
+                lhs.trim = rhs.trim;
+            }
+            return func;
+        }
+        Variant::TextLiteral(lhs) if let Variant::TextLiteral(rhs) = &mut *arg.variant => {
+            panic!("lhs: {:?}, rhs: {:?}", lhs, rhs)
+        }
```

Comment:
I didn't mean to merge this debug case. It only affects handling of a certain error; I will fix it in the next PR.